Adding Game Summary

VirusTLNR opened the pull request:
bugs include:-
1- sometimes the data is not loading in fuil
2- players who disconnect are not included in the summary sometimes?

otherwise data loads into a table at the end game, has player names, roles (w/color), modifiers (w/color), shows tasks counts, and shows who won.

the layout and ideas which lead to this working were pulled from TOR's coding, so should ref/credit whoever wrote the endgame summary section for TOR in regards to the layout of this.

ive scrapped the "rolehistory" addition.. because that was intereing with getting this working and so i figure ill add it in in a future commit.

VirusTLNR commented:
this is how it "should" show using the coding from the first commit.
![image](https://user-images.githubusercontent.com/80432742/145684794-12714f43-65d6-4d33-bb24-f00ccfea1836.png)

but for the host, the table does not load. (it was working before, but i made a change and reverted the change, and suddenly it stopped working.. i really should learn to not change things that work before backing them up :3
![image](https://user-images.githubusercontent.com/80432742/145684814-8e2a770f-a1ee-43e4-8654-0955fa51734d.png)

VirusTLNR commented:
oh and tasks are broken again.. (they were working too.. :/)
this is how it looked when it was working before i tried to improve some things, ill get looking into the bugs are occuring now when i get some time tomorrow.
![image](https://user-images.githubusercontent.com/80432742/145684873-d0decf0c-194a-4e14-9ff4-a94cda82abf9.png)

amsryq commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,218 @@
+        public static bool playerInfoUpdateFlag1Updating = false;
+        public static bool playerInfoUpdateFlag2EndOfGame = false;
+
+        public static void UpdatePlayerDataRPCCall()
```
Why is this RPC call needed?

VirusTLNR replied:
well this where i am baffled.

for just the end game summary, it could be called as the game ends, but that is not working, i placed it above the point where the winners setup, and it does not pull the data.

I was putting it in the tasks patch to be called whenever the task bar updates, and that works, but i think that is being called at the end of the game causing data to be only partial somehow? as i say, it was working fine in task patches until i changed something then it broke.

VirusTLNR replied:
i have fixed this, it now works within the task patch being run, this allows for constant updating meaning people can add additions (like task counts in meetings for dead people to see, etc etc) if it is desired, the problem was i put the assignment back to AdditionalTempData.PlayerData was being done within the loop, when it should have been after the loop, a misplace } on my part.

amsryq commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,218 @@
+
+        public static void UpdatePlayerInfo(List<PlayerControl> players)
+        {
+            List<AdditionalTempData.PlayerInfo> pinfolist = new List<AdditionalTempData.PlayerInfo>();
+            pinfolist.Clear();
```
Why `Clear()` when it is a newly created list?
```suggestion
            var pinfolist = new List<AdditionalTempData.PlayerInfo>();
```

VirusTLNR replied:
no thats fair, ive been juggling with various ways of doing things, i thought i cleaned up the code, but i guess i left that in ;)

amsryq commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,218 @@
+                {
+                    var firstText = task.name;//.Cast<ImportantTextTask>();
+                    PluginSingleton<TownOfUs>.Instance.Log.LogMessage($"TaskName:- " + task.name);
+                    var modname = "junktonull";
+                    var rolename = "junktonull";
+                    if (!IsNull(pinfo, "Mod"))
+                        modname = modifier.Name;
+
+                    if (!IsNull(pinfo, "Role"))
+                        rolename = role.Name;
```
Where is this used?

VirusTLNR replied:
edit:- seems that i have left some coding in where it shouldent be, was moving stuff around, oops, will check this, thanks

amsryq commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,218 @@
+                pinfo.TasksCompleted = tasksdone;
+                pinfo.TasksTotal = taskstotal;
+                pinfolist.Add(pinfo);
+                tasksdone = 0;
+                taskstotal = 0;
```
Why is this reassigned to 0 and never used?

VirusTLNR replied:
coding i put in because i was getting treble task counts, but i realised i did something wrong and forgot to remove them. thanks for pointing this out.

amsryq commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,218 @@
+
+        public static bool HasTasks(Faction faction)
+        {
+            if (faction == Faction.Crewmates)
+            {
+                return true;
+            }
+            else
+            {
+                return false;
+            }
```
```suggestion
            return faction == Faction.Crewmates;
```
or
```suggestion
            return faction is Faction.Crewmates;
```

amsryq commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,218 @@
+
+        static bool IsNull(AdditionalTempData.PlayerInfo pinfo, string type)
+        {
+            if (type == "Role")
+            {
+                if (pinfo.CurrentRole != null)
+                {
+                    return false;
+                }
+                else
+                {
+                    return true;
+                }
+            }
+            else if (type == "Mod")
+            {
+                if (pinfo.Modifier != null)
+                {
+                    return false;
+                }
+                else
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                return true;
+            }
```
Why not use switch case?
```suggestion
            return type switch {
                "Role" => pinfo.CurrentRole == null,
                "Mod" => pinfo.Modifier == null,
                _ => true
            };
```

amsryq commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,218 @@
+                        won = $"[<color=#FAD934FF>W</color>]";
+                    }
+                }
+                var nameinfo = "";
+
+                nameinfo = $"{data.PlayerName}";
```
```suggestion
                var nameinfo = data.PlayerName;
```

amsryq commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,218 @@
+                nameinfo = $"{data.PlayerName}";
+
+
+                var taskInfo = "";//GameData.Instance.AllPlayers[data.PlayerId].PlayerName;
+                                  //if (HasTasks(data.Faction))
+                                  //{
+                taskInfo = data.Faction > 0 ? $"" : $" <color=#FAD934FF>({data.TasksCompleted}/{data.TasksTotal})</color>";
+                //}
```
```cs
public static bool HasTasks(Faction faction)
```
What is this for, then?
```suggestion
                var taskInfo = !HasTasks(data.Faction) ? "" : $" <color=#FAD934FF>({data.TasksCompleted}/{data.TasksTotal})";
```

VirusTLNR replied:
well.. this will be just me borking up my own code ;)

i was basically having multiple issues so commented out the function to try something else, which then worked, so i left the function commented out. thank you for correcting.

VirusTLNR replied:
your commit suggestion actually removed the </color> required at the end of the line, i will re-add this in with another commit.

amsyarasyiq reviewed (changes requested):
The RPC call is redundant, you can remove it and update AdditionalTempData.playerData before `AmongUsClient.OnGameEnd` is called instead.

amsryq commented:
> players who disconnect are not included in the summary sometimes

They should not. Disconnected players are removed from `PlayerControl.AllPlayerControls`. As role summary are done by checking it, disconnected players will not get listed. Another way around is to use `GameData.Instance.AllPlayers` and retrieve player data from there.

VirusTLNR reviewed (commented):
thanks for the assistance amsyarasyiq, some good streamlining, cleanup, changes,etc :)

VirusTLNR commented:
> The RPC call is redundant, you can remove it and update AdditionalTempData.playerData before `AmongUsClient.OnGameEnd` is called instead.

well, this is how i started doing it, but that was not working, i will look at this again, but the idea of doing an RPC call was that it could be called at any time to update the data, and the data then could be used for multiple purposes throughout the game.

VirusTLNR commented:
> > players who disconnect are not included in the summary sometimes
> 
> They should not. Players that are disconnected is removed from `PlayerControl.AllPlayerControls`. As role summary are done by checking it, disconnected players will not get listed. Another way around is to use `GameData.Instance.AllPlayers` and retrieve player data from there.

thank you, will give this a try, this is something i am lacking in (understanding of the different places data can come from).

VirusTLNR commented:
will submit a commit in a second that changes from using PlayerControl.AllPlayerControls to GameData.Instance.AllPlayers
 
i seem to have fixed all the issues but left some redundant code in for this commit that i will remove in a second commit.

thank you for the helpful suggestions and changes @amsyarasyiq

the below result was.. kill the crew, report the body, everyone skip, after meeting phantom did 1 task.. then the impostor disconnected.
![image](https://user-images.githubusercontent.com/80432742/145737472-9ebea2dd-0b2a-4d43-a17b-5a54e00e8052.png)

amsryq commented on source/Patches/CustomRPC.cs:
```diff
@@ -99,6 +99,7 @@ public enum CustomRPC
 
-        AddMayorVoteBank
+        AddMayorVoteBank,
+        UpdateGamePlayerControlData
```
You can remove this now.

VirusTLNR replied:
Forgot that line ;) thanks, will remove it tomorrow.

Nevermind, removed it now, didn't realise I can just edit it here.

VirusTLNR commented:
i marked this as ready for review because it works as intended without bugs, 

if there are improvements to the code that can be made, let me know and i will make the changes, i have checked the code again today and found no unnecessary lines that should be removed which were added by me. so wanted to confirm why i marked it ready for review initially, there was just one line i forgot about which is why there was an extra commit after the last review..

Thanks again for the assistance provided.

![image](https://user-images.githubusercontent.com/80432742/145851867-f3fe275f-7022-4777-9775-ec2a4fcfdd98.png)

amsyarasyiq reviewed (changes requested):
>  the data then could be used for multiple purposes throughout the game.

If so, move `AdditionalTempData` to somewhere outside `EndGameSummary` together with `UpdatePlayerInfo()`.

VirusTLNR commented:
> > the data then could be used for multiple purposes throughout the game.
> 
> If so, move `AdditionalTempData` to somewhere outside `EndGameSummary` together with `UpdatePlayerInfo()`.

if there was a multiple use of it i would, but i figured as the only use is the summary for now, i would leave it all within the summary. If you want to suggest a better place for it to go, i can move it though.

amsryq commented:
You may want to link issue #66 to this pull request.

Anusien commented:
Very cool.

Anusien commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,141 @@
+    {
+        static class AdditionalTempData
+        {
+            public static List<PlayerInfo> playerData = new List<PlayerInfo>();
```
Nit: can be `readonly` right?

VirusTLNR replied:
not possible because it gets assigned to on line 63.

Anusien commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,141 @@
+
+        public static void UpdatePlayerInfo()
+        {
+            Il2CppSystem.Collections.Generic.List<GameData.PlayerInfo> players = GameData.Instance.AllPlayers;
```
Nit: I think we'd want to prefer the built-in C# List.

amsryq replied:
It was obtained from GameData.Instance.AllPlayers which is IL2CPP collections' List so that might require `ToArray()` (convert and cast it to built-in array type) and then `ToList()` (to built-in List collection).

This is unnecessary as it only was used once in the foreach loop. I'd recommend implicitly type it using `var` or directly reference it in the foreach loop (`foreach (GameData.PlayerInfo player in GameData.Instance.AllPlayers`).

VirusTLNR replied:
i tried to use the built in list, but it did not work, an error of cannot implicitly convert between the two types.

I also tried casting and that did not seem to work either, so i stuck to what was working, as ugly as it looks, i used the normal List<T> where possible.

VirusTLNR replied:
> It was obtained from GameData.Instance.AllPlayers which is IL2CPP collections' List so that might require `ToArray()` (convert and cast it to built-in array type) and then `ToList()` (to built-in List collection).
> 
> This is unnecessary as it only was used once in the foreach loop. I'd recommend implicitly type it using `var` or directly reference it in the foreach loop (`foreach (GameData.PlayerInfo player in GameData.Instance.AllPlayers`).

I was using toArray().toList() but it threw the same error as that .toList() is a system list and it didnt like the conversion.

It could be there is a way this works, but I tried a  lot of ways and nothing seemed to. I had this issue with both PlayerControl and GameData.Instance.AllPlayers

Anusien commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,141 @@
+        {
+            Il2CppSystem.Collections.Generic.List<GameData.PlayerInfo> players = GameData.Instance.AllPlayers;
+
+            var pinfolist = new List<AdditionalTempData.PlayerInfo>();
```
Nit: I prefer using fully qualified types over `var`; I find them much easier to read and reason about.

VirusTLNR replied:
i agree, makes it easier to read, but var makes changing code easier, as all the variables change with the assignment.

i have changed this to the fully qualified type as i dont expect this code to change anymore.

Anusien replied:
I don't think "speed of rewriting the code" is the important thing to optimize. Your code has to be read by a human many more times than it is modified by a human, so we should optimize for the thing which happens more often.

Anusien commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,141 @@
+            return faction == Faction.Crewmates;
+        }
+
+        static bool IsNull(AdditionalTempData.PlayerInfo pinfo, string type)
```
Suggestion: I think this method will cause more confusion in the long run than it will save you time here. Especially because "Role" and "Mod" are strings. For example, it's not obvious that you need to put "Mod" instead of "Modifier", and there's not really any feedback to know you've gotten it wrong.

VirusTLNR replied:
i am well aware i was doing an ugly way of doing this, but i found a big cause of problem for me was the null exceptions from modifier mostly, but role also. So once i got it working, i did not review it, after making the other changes you suggested, i will look at improving this.

VirusTLNR replied:
it seems !=null works now.. baffling, must have been something else in my code that was breaking it before, either way, simplified this now to if(variable!=null)

Anusien commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,141 @@
+
+            foreach (var data in AdditionalTempData.playerData)
+            {
+                var won = "   ";
```
Suggestion: Java has a keyword `final`; any variable defined as `final` can only be written to exactly one. C# doesn't actually have a `final`, but I suggest you write code that would work that way. So instead of using a `won` string and rewrite it, define the variable once and write to it once in each branch. So something like:

```string won;
if (TempData.winners.Contains(w -> w.Name == data.PlayerName)
{
   won = won = $"[<color=#FAD934FF>W</color>]";
}
else {
   won = "   ";
}
```

This is a pretty benign example, but it's a good habit to get into because re-assigning variables can make it *very* hard to understand what a piece of code does later when you read it.

VirusTLNR replied:
i have changed it so the branches are more clear as suggested, couldent get the -> working with WinningPlayerData though as in your suggested code, so did it the same way i was already doing it.

VirusTLNR replied:
i have had to revert the changes i made for this as the changes i made seem to be causing the bug with the lovers not being shown as winners... i will put a comment in to reflect this.

Anusien commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,141 @@
+                }
+                var nameinfo = data.PlayerName;
+
+                List<string> noncrewroleswithtasks = new List<string>();
```
Nit: Consider camelCase for variable names to make them easier to read.

VirusTLNR replied:
usually do, forgot it here, will edit when i get a chance.

Anusien commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,141 @@
+            AdditionalTempData.playerData = pinfolist;
+        }
+
+        public static bool HasTasks(Faction faction)
```
Suggestion: This is duplicating some code that actually determines whether the player can do tasks. Specifically it duplicates the code that blocks neutral players from doing tasks. Might want to just call that method.

VirusTLNR replied:
removed the requirement for this function by putting the bool statement in the line, to be honest not aware there is a function that calculates this elsewhere, i have looked in a lot of placed id expect to find it and could not see such a function, i maybe just missing it though.

Anusien replied:
I think we should create such a helper method in `TasksPatch.cs`; see my other comment for why.

Anusien commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,141 @@
+                        taskstotal++;
+                }
+                
+                pinfo.TasksCompleted = tasksdone;
```
Suggestion: I think you can get rid of the intervening variables and just write to this directly.

VirusTLNR replied:
indeed, and done.

Anusien commented on source/Patches/Roles/Modifiers/Modifier.cs:
```diff
@@ -65,6 +65,12 @@ public static Modifier GetModifier(PlayerControl player)
 
+        public static Modifier GetModifier(GameData.PlayerInfo player)
+        {
+            return (from entry in ModifierDictionary where entry.Key == player.PlayerId select entry.Value)
```
Possible bug: Do we need a check for `player == null` the way we do in `Role#GetRole(GameData.PlayerInfo)`?

VirusTLNR replied:
not going to lie, i just straight up replicated what i did in Role... and applied it to Modifier, the method in modifier did not do the null check, so i left it out, might be something that needs fixing in multiple places, as a big part of the problem i had with modifiers was that they would be null and stop everything from working, leading to me writing a method for checking if its null which probably isnt needed.

VirusTLNR replied:
so    public static Modifier GetModifier(PlayerControl player) will need changing too if thats the case.

Anusien replied:
The code for `GetRole(GameData.PlayerInfo)` does an explicit check for `player == null` but this code doesn't. This code seems to do something very similar to that code, but it looks *totally* different. Seems like they should be very similar unless there's a compelling reason not to.

VirusTLNR replied:
i have put if (player == null) return null; into both the old function, and the new one i added in modifier.cs (the two listed below) and all seems to be working.

public static Modifier GetModifier(PlayerControl player)
public static Modifier GetModifier(GameData.PlayerInfo player)

Anusien commented:
> > > the data then could be used for multiple purposes throughout the game.
> > 
> > 
> > If so, move `AdditionalTempData` to somewhere outside `EndGameSummary` together with `UpdatePlayerInfo()`.
> 
> if there was a multiple use of it i would, but i figured as the only use is the summary for now, i would leave it all within the summary. If you want to suggest a better place for it to go, i can move it though.

https://github.com/Anusien/Town-Of-Us/blob/fcfc866494edb89f628b58c3e44c1ed350f2557f/source/Patches/TaskPatches.cs#L24

Looks like it's now just looking at `faction == Faction.Crewmates` (it didn't use to, but I fixed it). Still I think we ought to refactor this to a helper method because other mods have roles with more complicated logic, so it's not out of the question that we should here.

Anusien commented on source/Patches/TaskPatches.cs:
```diff
@@ -28,6 +28,7 @@ private static bool Prefix(GameData __instance)
                             if (playerInfo.Tasks.ToArray()[j].Complete) __instance.CompletedTasks++;
                         }
                 }
+                Patches.EndGameSummary.UpdatePlayerInfo();
```
Is there a reason we can't just do this right before we need it? If so, would love to have that documented in the code. Comments that explain "why" we do something, especially something weird or counter-intuitive, are very important.

VirusTLNR replied:
i will try again putting it just before the end game, but it wasnt working before, so i had to put it back in TaskPatches,

ive put it back at the place i did before (AmongUsClient_OnGameEnd.PostFix) and it works there again, so i will remove it from TaskPatches.

regarding comments, i normally put in tonnes of comments, i hate uncommented code, but i saw no comments in the code in other PRs so figured the comments were not wanted and removed them before i committed.

VirusTLNR commented:
just found a bug while changing the code.. lovers dont both get shown as winners, this may have been working before, but im in the middle of changing the code and suddenly it doesnt work.

![image](https://user-images.githubusercontent.com/80432742/146100765-00af7ec7-1ce5-43ea-977e-092acf35e902.png)

Anusien commented:
> just found a bug while changing the code.. lovers dont both get shown as winners, this may have been working before, but im in the middle of changing the code and suddenly it doesnt work.
> 
> ![image](https://user-images.githubusercontent.com/80432742/146100765-00af7ec7-1ce5-43ea-977e-092acf35e902.png)

If you have trouble getting the lovers winning to show up right, just remove indication of who won (people can easily tell since the winners are right in front of them) and file an issue. I’d rather deploy this without it rather than spend several days investigating a minor issue.

VirusTLNR commented:
> > just found a bug while changing the code.. lovers dont both get shown as winners, this may have been working before, but im in the middle of changing the code and suddenly it doesnt work.
> > ![image](https://user-images.githubusercontent.com/80432742/146100765-00af7ec7-1ce5-43ea-977e-092acf35e902.png)
> 
> If you have trouble getting the lovers winning to show up right, just remove indication of who won (people can easily tell since the winners are right in front of them) and file an issue. I’d rather deploy this without it rather than spend several days investigating a minor issue.

i found the issue, it was the in progress code changed for setting an if and else statement, run 3/4 games and both lovers show as winners in all of them, will public the commit in a second.
![image](https://user-images.githubusercontent.com/80432742/146104820-cc81ca79-53a3-4a32-a0f6-3ac2f7551520.png)

i was getting this as an output when i was getting only 1 lover being shown as a winner. (true = W should be shown, but only the second one showed every time)
[Message:Town Of Us] Lover:- Virus
[Message:Town Of Us] Role:- Lover
[Message:Town Of Us] Lover:- Virus 1
[Message:Town Of Us] Role:- Lover
[Message:Town Of Us] WinnerName=#Virus# vs PlayerName=#Virus#
[Message:Town Of Us] TRUE
[Message:Town Of Us] WinnerName=#Virus 1# vs PlayerName=#Virus#
[Message:Town Of Us] FALSE
[Message:Town Of Us] WinnerName=#Virus# vs PlayerName=#Virus 1#
[Message:Town Of Us] FALSE
[Message:Town Of Us] WinnerName=#Virus 1# vs PlayerName=#Virus 1#
[Message:Town Of Us] TRUE
[Message:Town Of Us] WinnerName=#Virus# vs PlayerName=#Virus 2#
[Message:Town Of Us] FALSE
[Message:Town Of Us] WinnerName=#Virus 1# vs PlayerName=#Virus 2#
[Message:Town Of Us] FALSE
[Message:Town Of Us] WinnerName=#Virus# vs PlayerName=#Virus 3#
[Message:Town Of Us] FALSE
[Message:Town Of Us] WinnerName=#Virus 1# vs PlayerName=#Virus 3#
[Message:Town Of Us] FALSE

VirusTLNR commented:
that commit is everything working as i said, thanks for the assistance btw :)

amsryq commented on source/Patches/EndGameSummary.cs:
```diff
@@ -0,0 +1,141 @@
+                foreach (WinningPlayerData winner in TempData.winners)
+                {
+                    Reactor.PluginSingleton<TownOfUs>.Instance.Log.LogMessage($"WinnerName=#{winner.Name}# vs PlayerName=#{data.PlayerName}#");
+                    if (winner.Name == data.PlayerName)
+                    {
+                        Reactor.PluginSingleton<TownOfUs>.Instance.Log.LogMessage($"TRUE");
+                        won = $"[<color=#FAD934FF>W</color>]";
+                    }
```
I believe the issue was you forgetting to add `break;` causing the loop to continue (and assigns `won` as a blank white spaces).

VirusTLNR replied:
yes, apparantly i forgot a break after the won, this is why i put it as just if() without an else before, as now it re-assigns blank spaces until it finds the right player, i have changed this though, thanks for pointing this out.

VirusTLNR commented:
bug now fixed regarding the winners W, also added in a shuffle() to re-order the players in the summary list so people cant identify the exact player join order to use as info in the games. (a streamer used this in the past)

Also moved the list into the loop creation as suggested so no more ugly one off reference to the other list type.
![image](https://user-images.githubusercontent.com/80432742/146200573-dc48574e-f2bb-479b-abab-02d6c8c22dd7.png)

Anusien commented:
Looks good to me, thanks for all the hard work and your continued refinements.

I'll try to cut a release with this sometime today.
